### PR TITLE
lite framework prototype

### PR DESCRIPTION
**Pull request by Superjomn**
A prototype for multi-hardware support, high performance underlying computation framework.

The core concepts:

- `OpLite`: the light-weight Operator interface
- `KernelLite`: the kernel with better hardware support
- `TypeSystem`: the type system to make the type inference across the whole computation graph
- `MIR`: the machine IR, the underlying IR and Passes with runtime information concerned

In the new framework, we concentrate on the following aspects:

1. separation of  the analysis phase and the execution phase, which improves the runtime performance,
2. strong type inference to make the runtime system more visible,
3. the underlying analysis and optimization based on MIR, uncoupled from the upper analysis, transferable on multiple hardware.

The new architecture:

image

# Compile

For server side:

GPU:

```bash
make .. -DWITH_GPU=ON  -DWITH_LITE=ON -DLITE_WITH_LIGHT_WEIGHT_FRAMEWORK=OFF -DLITE_WITH_CUDA=ON -DWITH_TESTING=ON
make test_cxx_api_lite -j8
```
CPU:

```bash
make .. -DWITH_GPU=OFF  -DWITH_LITE=ON -DLITE_WITH_LIGHT_WEIGHT_FRAMEWORK=OFF -DLITE_WITH_CUDA=OFF -DWITH_TESTING=ON
make test_cxx_api_lite -j8
````

For mobile side:

```bash
make .. -DWITH_GPU=OFF -DWITH_LITE=ON -DLITE_WITH_LIGHT_WEIGHT_FRAMEWORK=ON -DLITE_WITH_CUDA=OFF -DWITH_TESTING=ON
make test_cxx_api_lite -j8
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -79,19 +79,55 @@ option(PY_VERSION       "Compile PaddlePaddle with python3 support"     ${PY_VER
...
-# PY_VERSION
-if(NOT PY_VERSION)
-  set(PY_VERSION 2.7)
-endif()
-set(PYBIND11_PYTHON_VERSION ${PY_VERSION})
+# for lite, both server and mobile framework.
+option(WITH_LITE "Enable lite framework" OFF)
+option(LITE_WITH_CUDA "Enable CUDA in lite mode" OFF)
+option(LITE_WITH_X86  "Enable X86 in lite mode" ON)
+option(LITE_WITH_LIGHT_WEIGHT_FRAMEWORK  "Enable light-weight framework" OFF)
```
Comment (luotao1) [inline CMakeLists.txt:86]: light-weight framework和lite framework有什么区别呢？注释能否更清晰一点。
`LITE_WITH_LIGHT_WEIGHT_FRAMEWORK`这个是指mobile么？这个选项改成`LITE_WITH_MOBILE`?是否更清晰

File: CMakeLists.txt [hunk 2]
```diff
@@ -79,19 +79,55 @@ option(PY_VERSION       "Compile PaddlePaddle with python3 support"     ${PY_VER
...
+    include(external/protobuf)  # download, build, install protobuf
+    include(external/eigen)     # download eigen3
+
+    include(generic)            # simplify cmake module
+    include(configure)          # add paddle env configuration
+
+    add_definitions(-std=c++11)
+
+    add_subdirectory(paddle)
+
```
Comment (luotao1) [inline CMakeLists.txt:120]: 可去掉113，116，118，120多余的空行。

File: cmake/configure.cmake [hunk 3]
```diff
@@ -157,3 +157,21 @@ endif(WITH_BRPC_RDMA)
 if(ON_INFER)
     add_definitions(-DPADDLE_ON_INFERENCE)
 endif(ON_INFER)
+
+if(WITH_WBAES)
+    add_definitions(-DPADDLE_WITH_WBAES)
+endif(WITH_WBAES)
```
Comment (luotao1) [inline cmake/configure.cmake:163]: 请问`WITH_WBAES`是用在哪里呢？没有看到这个选项。

File: paddle/CMakeLists.txt [hunk 4]
```diff
@@ -1,4 +1,7 @@
-add_subdirectory(scripts)
-add_subdirectory(testing)
-set(PYTHON_TESTS_DIR ${PADDLE_BINARY_DIR}/python/paddle/fluid/tests CACHE INTERNAL "python tests directory")
+# to limit the mobile dependencies
+if (NOT LITE_WITH_LIGHT_WEIGHT_FRAMEWORK)
+  add_subdirectory(scripts)
+  add_subdirectory(testing)
+  set(PYTHON_TESTS_DIR ${PADDLE_BINARY_DIR}/python/paddle/fluid/tests CACHE INTERNAL "python tests directory")
+endif()
 add_subdirectory(fluid)
```
Comment (luotao1) [inline paddle/CMakeLists.txt:7]:
```
add_subdirectory(fluid)
if (NOT LITE_WITH_LIGHT_WEIGHT_FRAMEWORK)
  return()
endif()
add_subdirectory(scripts)
add_subdirectory(testing)
set(PYTHON_TESTS_DIR ${PADDLE_BINARY_DIR}/python/paddle/fluid/tests CACHE INTERNAL "python tests directory")
```
这样更加清晰。

File: paddle/fluid/CMakeLists.txt [hunk 5]
```diff
@@ -6,7 +10,8 @@ add_subdirectory(operators)
 add_subdirectory(string)
 add_subdirectory(recordio)
 add_subdirectory(pybind)
-
+add_subdirectory(train)
 # NOTE: please add subdirectory inference at last.
 add_subdirectory(inference)
-add_subdirectory(train)
+
+add_subdirectory(lite)
```
Comment (luotao1) [inline paddle/fluid/CMakeLists.txt]: 17行重复加了`add_subdirectory(lite)`？
是想表达lite这个目录一直都要加么？下面改法可能更为清晰。
```
add_subdirectory(lite)
if (LITE_WITH_LIGHT_WEIGHT_FRAMEWORK)
  return()
endif()
add_subdirectory(memory)
...
```

File: paddle/fluid/framework/CMakeLists.txt [hunk 6]
```diff
@@ -124,7 +124,7 @@ cc_library(shape_inference SRCS shape_inference.cc DEPS ddim attribute device_co
 cc_library(transfer_scope_cache SRCS transfer_scope_cache.cc DEPS scope framework_proto device_context)
 cc_library(op_kernel_type SRCS op_kernel_type.cc DEPS device_context place)
 cc_library(operator SRCS operator.cc DEPS op_info device_context tensor scope glog
-    shape_inference data_transform lod_tensor profiler transfer_scope_cache op_kernel_type)
+    shape_inference data_transform lod_tensor profiler transfer_scope_cache op_kernel_type data_feed_proto)
```
Comment (luotao1) [inline paddle/fluid/framework/CMakeLists.txt:127]: data_feed_proto是从哪儿来的？

File: CMakeLists.txt (hunk 1 above)
Comment (Superjomn) [reply CMakeLists.txt:86]: 暂时不会暴露 mobile 的信息，轻量级框架不分平台和硬件

File: paddle/fluid/framework/CMakeLists.txt (hunk 6 above)
Comment (Superjomn) [reply paddle/fluid/framework/CMakeLists.txt:127]: 这个是framework 里的依赖有问题，加上才能正常编译

**Review by XiaoguangHu01 — DISMISSED** [review]

**Review by jacquesqiao — CHANGES_REQUESTED** [review]
这个pr还是慎重点吧..

**Comment by Superjomn** [comment]
这个一阶段是预测的迁移 @jacquesqiao

**Review by jacquesqiao — APPROVED** [review]
这个操作只是为了撤销 request change

File: paddle/fluid/framework/framework.proto [hunk 7]
```diff
@@ -13,7 +13,7 @@ See the License for the specific language governing permissions and
 limitations under the License. */
 
 syntax = "proto2";
-option optimize_for = LITE_RUNTIME;
+//option optimize_for = LITE_RUNTIME;
```
Comment (jacquesqiao) [inline paddle/fluid/framework/framework.proto:16]: what is the purpose of this line?

File: paddle/fluid/framework/op_desc.h [hunk 8]
```diff
@@ -33,7 +33,7 @@ class OpDesc {
   OpDesc(const std::string &type, const VariableNameMap &inputs,
          const VariableNameMap &outputs, const AttributeMap &attrs);
 
-  OpDesc(const proto::OpDesc &desc, BlockDesc *block);
+  explicit OpDesc(const proto::OpDesc &desc, BlockDesc *block = nullptr);
```
Comment (jacquesqiao) [inline paddle/fluid/framework/op_desc.h:36]: only constructor with one parameter need to add `explicit`

**Comment by Superjomn** [comment]
More validation will be done later, close temporarily.

**Pull request closed without merging**